west 0.10.0 and 0.9.1 docs

mbolivar-nordic opened the pull request:
This PR documents new `group-filter` semantics proposed for west 0.10 in https://github.com/zephyrproject-rtos/west/pull/482 along with changes for v0.9.1.

DNM because it can't be merged until we've cut these two releases.

marc-hb committed:
doc: west: submodule 'name' is optional after west 0.9.0

See west commit 02444d058335

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

mbolivar-nordic committed:
doc: west: move 'manifest imports' to bottom of manifest.rst

This diff is deceptively complicated. No text is being added or
removed; the only thing that's being done is that the entire 'Manifest
Imports' section is being moved down below the groups and submodules
sections.

This is prep work for a later commit which will document revised
semantics for mixing 'import' with 'group-filter'. This will make
'group-filter' values compose across imports, so it will be useful to
have documentation for group and group-filter before documentation for
imports. That will let us document the way they work together within
the 'Manifest Imports' section without using concepts that haven't yet
been defined.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: fix incorrect docs related to imports

The documentation suggests that 'west update PROJECT' can be done if
PROJECT is defined from some imported manifest. That's actually not
true. Fix it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: clean up manifest import details

Move the content related to handling imports that refer to multiple
files handled to right after the ordered list which defines the order
that manifests are imported. The current placement at the end of this
docs section was poorly chosen.

Create subsections for how projects and extensions are decided in the
final resolved manifest.

The project revisions in the top level example manifest fragment are
not important here. Remove them for clarity.

The top level manifest file need not be named west.yml now; adjust
accordingly.

This is prep work for adding a new subsection for group-filter.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: document west 0.10 group-filter + import behavior

The west 0.9 documentation for group-filter states:

    Only the top level manifest file’s manifest: group-filter:
    value has any effect. The manifest: group-filter: values in any
    imported manifests are ignored.

This turns out to have been a mistake, because all users who import a
manifest which makes some projects inactive will have to also manually
disable the same groups just to get the same default list of projects.

Example manifest fragments showing the issue:

   # my/west.yml
   manifest:
     projects:
       - name: upstream
         import: true

and:

   # upstream/west.yml
   manifest:
     group-filter: [-disabled-group]
     projects:
       - name: i-want-this
       - name: i-do-not-want-this
         groups:
         - disabled-group

As written, my/west.yml will include 'i-do-not-want-this' as an
active project.

This kind of leaky faucet has proven to be an unintuitive and poor
user experience. Users expect to get the default projects without
having to copy the group-filter for everything they import.

We can't reverse this behavior without pushing out a new schema
version, however: manifest schema 0.9, and therefore west 0.9.x, is
doomed to this bad behavior.

What we'll do to move past this is create a new schema version 0.10
that does the right thing, document the 0.10 behavior, and discourage
combining these two features in west 0.9.x.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -788,6 +787,120 @@ In this case, both ``groupA`` and ``groupB`` are disabled.
+
+.. warning::
+
+   The semantics are different in in west v0.9.x. There, only the top level
```
```suggestion
   The semantics are different in west v0.9.x. There, only the top level 
```

tejlmand commented on doc/guides/west/manifest.rst:
```diff
@@ -788,6 +787,120 @@ In this case, both ``groupA`` and ``groupB`` are disabled.
+   manifest file's ``group-filter`` had any effect. This was changed in west
+   0.10.0 to the behavior described below.
+
+   Combining ``group-filter`` with ``import`` in west v0.9.x is discouraged.
```
Maybe simplify this.
Not sure readers care so much about the details, but should simply be warned about west `0.9.0` and import.

Just like we don't warn about the west 0.7.x import and missing `self:path` in manifest (I helped someone on slack just yesterday that faced this issue)

```suggestion
   In west v0.9.x only the top level ``group-filter`` has effect, therefore
   combining ``group-filter`` with ``import`` in west v0.9.x is discouraged and west v0.10.0 or newer should be used.
```

mbolivar-nordic replied:
Changed to:

```
   The below semantics apply to west v0.10.0 and later. West v0.9.x semantics
   are different, and combining ``group-filter`` with ``import`` in west v0.9.x
   is discouraged.
```

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -788,6 +787,118 @@ In this case, both ``groupA`` and ``groupB`` are disabled.
+
+The same :file:`parent/west.yml` file disables the ``optional`` group, so
+``project-2`` is inactive.
+
```
Great section! The only thing I am missing in it is perhaps a mention to how `west config manifest.group-filter` can modify this. You did mention this in the beginning of the section, but perhaps a third example that replicates the second one while also adding a call to `west config manifest.group-filter +optional` so that we show all projects enabled?

mbolivar-nordic replied:
>  perhaps a third example

Good idea. Now added.

carlescufi commented on doc/guides/west/manifest.rst:
```diff
@@ -1757,6 +1871,31 @@ If an imported manifest file has a ``west-commands:`` definition in its
+
+The final resolved ``group-filter`` value is then ``filter1 + filter-2 + ... +
+filter-N + top-filter + self-filter``, where ``+`` here refers to list
+concatenation.
```
I am missing something here. From the example in the "Group Filters and Imports" section, the sequence would be:
`group-filter = [-unstable] +  [+unstable,-optional]`, which leads to `[+unstable, -optional]`.

But if the order would have been:
`group-filter = [+unstable] +  [-unstable,-optional]` then the end result would have had to be `[-unstable,-optional]`.
Is this captured in the resolution sequence of additions above?

carlescufi replied:
i.e. this is not a standard commutative sum, which I am not sure is well reflected.

mbolivar-nordic replied:
> i.e. this is not a standard commutative sum, which I am not sure is well reflected.

I've tried to address this.

carlescufi reviewed (commented):
I have a suggestion and a question, but overall this looks great!

mbolivar-nordic committed:
doc: west: projects list is optional in v0.10

Document this fact.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: touch up remotes documentation

Try to make it a bit clearer.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: bump 'manifest: version:' docs

Adjust these for 0.10. This version parses differently as a string vs.
as a float. Document that quoting the value avoids the issue. Make
some other adjustments and improvements for clarity.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: 0.10.0 and 0.9.1 release notes

Combining these since they are related.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
I've bundled the docs for 0.10.0 and 0.9.1 here after cherry-picking https://github.com/zephyrproject-rtos/zephyr/pull/32106 (for which cc @marc-hb).

Ready for final review. DNM until we have cut both releases.

tejlmand reviewed (commented):
Knowing how fond i'm of copy-paste, should we also update this line:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/guides/west/manifest.rst#L392

to:
```
manifest:
  version: '0.10'
```

and based on my observations here:
https://github.com/zephyrproject-rtos/west/pull/482#pullrequestreview-607532039
also quote it ?

mbolivar-nordic commented:
Good point @tejlmand. I've reworked the section about manifest versions completely to try to improve this situation. Please take another look.

mbolivar-nordic committed:
.github: doc-build: use west 0.10.0a1

Testing for the release itself.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand reviewed (approved):
LGTM.

Thanks for improving an already great doc even more.
